manager/manager.go

```diff
@@ -1096,7 +1096,11 @@ func (m *manager) getContainersDiff(containerName string) (added []info.Containe
 	if !ok {
 		return nil, nil, fmt.Errorf("failed to find container %q while checking for new containers", containerName)
 	}
-	allContainers, err := cont.handler.ListContainers(container.ListRecursive)
+	contHandler := cont.handler
+	m.containersLock.RUnlock()
```

Comment:
nit: rather than locking for all but the `ListContainers` call, can we instead lock around just the critical sections?  I.E. lock and unlock around the `:= m.containers` call above, and then Lock, and defer Unlock before the `Determine which were added... ` section